src/main/drivers/serial_uart_stm32h7xx.c

```diff
@@ -0,0 +1,567 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * jflyper - Refactoring, cleanup and made pin-configurable
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "platform.h"
+
+#ifdef USE_UART
+
+#include "drivers/system.h"
+#include "drivers/dma.h"
+#include "drivers/io.h"
+#include "drivers/nvic.h"
+#include "drivers/rcc.h"
+
+#include "drivers/serial.h"
+#include "drivers/serial_uart.h"
+#include "drivers/serial_uart_impl.h"
+
+#ifndef UART1_TX_DMA_STREAM
+#define UART1_TX_DMA_STREAM NULL
+#endif
+#ifndef UART1_RX_DMA_STREAM
+#define UART1_RX_DMA_STREAM NULL
+#endif
+#ifndef UART2_TX_DMA_STREAM
+#define UART2_TX_DMA_STREAM NULL
+#endif
+#ifndef UART2_RX_DMA_STREAM
+#define UART2_RX_DMA_STREAM NULL
+#endif
+#ifndef UART3_TX_DMA_STREAM
+#define UART3_TX_DMA_STREAM NULL
+#endif
+#ifndef UART3_RX_DMA_STREAM
+#define UART3_RX_DMA_STREAM NULL
+#endif
+#ifndef UART4_TX_DMA_STREAM
+#define UART4_TX_DMA_STREAM NULL
+#endif
+#ifndef UART4_RX_DMA_STREAM
+#define UART4_RX_DMA_STREAM NULL
+#endif
+#ifndef UART5_TX_DMA_STREAM
+#define UART5_TX_DMA_STREAM NULL
+#endif
+#ifndef UART5_RX_DMA_STREAM
+#define UART5_RX_DMA_STREAM NULL
+#endif
+#ifndef UART6_TX_DMA_STREAM
+#define UART6_TX_DMA_STREAM NULL
+#endif
+#ifndef UART6_RX_DMA_STREAM
+#define UART6_RX_DMA_STREAM NULL
+#endif
+#ifndef UART7_TX_DMA_STREAM
+#define UART7_TX_DMA_STREAM NULL
+#endif
+#ifndef UART7_RX_DMA_STREAM
+#define UART7_RX_DMA_STREAM NULL
+#endif
+#ifndef UART8_TX_DMA_STREAM
+#define UART8_TX_DMA_STREAM NULL
+#endif
+#ifndef UART8_RX_DMA_STREAM
+#define UART8_RX_DMA_STREAM NULL
+#endif
+
+#ifdef USE_UART1
+DMA_RAM volatile uint8_t uart1RxBuffer[UART_RX_BUFFER_SIZE];
+DMA_RAM volatile uint8_t uart1TxBuffer[UART_TX_BUFFER_SIZE];
+#endif
+
+#ifdef USE_UART2
+DMA_RAM volatile uint8_t uart2RxBuffer[UART_RX_BUFFER_SIZE];
+DMA_RAM volatile uint8_t uart2TxBuffer[UART_TX_BUFFER_SIZE];
+#endif
+
+#ifdef USE_UART3
+DMA_RAM volatile uint8_t uart3RxBuffer[UART_RX_BUFFER_SIZE];
+DMA_RAM volatile uint8_t uart3TxBuffer[UART_TX_BUFFER_SIZE];
+#endif
+
+#ifdef USE_UART4
+DMA_RAM volatile uint8_t uart4RxBuffer[UART_RX_BUFFER_SIZE];
+DMA_RAM volatile uint8_t uart4TxBuffer[UART_TX_BUFFER_SIZE];
+#endif
+
+#ifdef USE_UART5
+DMA_RAM volatile uint8_t uart5RxBuffer[UART_RX_BUFFER_SIZE];
+DMA_RAM volatile uint8_t uart5TxBuffer[UART_TX_BUFFER_SIZE];
+#endif
+
+#ifdef USE_UART6
+DMA_RAM volatile uint8_t uart6RxBuffer[UART_RX_BUFFER_SIZE];
+DMA_RAM volatile uint8_t uart6TxBuffer[UART_TX_BUFFER_SIZE];
+#endif
+
+#ifdef USE_UART7
+DMA_RAM volatile uint8_t uart7RxBuffer[UART_RX_BUFFER_SIZE];
+DMA_RAM volatile uint8_t uart7TxBuffer[UART_TX_BUFFER_SIZE];
+#endif
+
+#ifdef USE_UART8
+DMA_RAM volatile uint8_t uart8RxBuffer[UART_RX_BUFFER_SIZE];
+DMA_RAM volatile uint8_t uart8TxBuffer[UART_TX_BUFFER_SIZE];
+#endif
+
+const uartHardware_t uartHardware[UARTDEV_COUNT] = {
+#ifdef USE_UART1
+    {
```

Comment:
I'd use macro here too, it will considerably reduce code duplication ... 
Some preprocessor magic may be necessary here to expand embedded macros correctly